File: threads_attack_solution.c
#include <pthread.h>
#include <stdbool.h>
#include <stdio.h>

const int NUM_THREADS = 2;
const int NUM_ITERS = 1000000;

typedef struct {
  _Atomic bool wantsToEnter1;
  _Atomic bool wantsToEnter2;
  _Atomic int turn;
}  MyLock;

MyLock buildLock() {
  return ((MyLock) {
    .wantsToEnter1 = false,
    .wantsToEnter2 = false,
    .turn = 1
  });
}

typedef struct {
  MyLock* lock;
  int* counter;
  int threadIdx;
} ThreadArgument;

void performLock(MyLock* lock, int threadIdx) {
  if (threadIdx == 1) {
    lock->turn = 2;
    lock->wantsToEnter1 = true;
    while (lock->wantsToEnter2 && (lock->turn == 2)) {
      // Just busy wait.
    } 
  } else {
    lock->turn = 1;
    lock->wantsToEnter2 = true;
    while (lock->wantsToEnter1 && (lock->turn == 1)) {
      // Just busy wait.
    } 
  }
}

void performUnlock(MyLock* lock, int threadIdx) {
  if (threadIdx == 1) {
    lock->wantsToEnter1 = false;
  } else {
    lock->wantsToEnter2 = false;
  }
}

void* thread_main(void* argument) {
  ThreadArgument* threadArgument = (ThreadArgument*) argument;

  for (int i = 0; i < NUM_ITERS; i ++) {
    performLock(threadArgument->lock, threadArgument->threadIdx);
    *(threadArgument->counter) += 1;
    performUnlock(threadArgument->lock, threadArgument->threadIdx);
  }

  return NULL;
}

int main() {
  int counter = 0;
  MyLock lock = buildLock();
  
  ThreadArgument threadArgument1 = ((ThreadArgument) {
    .lock = &lock,
    .counter = &counter,
    .threadIdx = 1
  });
  ThreadArgument threadArgument2 = ((ThreadArgument) {
    .lock = &lock,
    .counter = &counter,
    .threadIdx = 2
  });
  
  pthread_t threads[NUM_THREADS];

  pthread_create(&threads[0], NULL, thread_main, &threadArgument1);
  pthread_create(&threads[1], NULL, thread_main, &threadArgument2);
  
  for (int i = 0; i < NUM_THREADS; i ++) {
    pthread_join(threads[i], NULL);
  }
  
  printf("Final Result: %d", counter);
  
  return 0;
}

File: threads_attack_solution.rb
require 'concurrent'
require 'thread'

NUM_TIMES = 1_000_000

class MyLock
  def initialize
    @wants_to_enter1 = Concurrent::AtomicBoolean.new(false)
    @wants_to_enter2 = Concurrent::AtomicBoolean.new(false)
    
    @turn = 1
  end

  def lock1
    @wants_to_enter1.make_true
    @turn = 2
    while @wants_to_enter2.true? && (@turn == 2)
      # Busy wait
    end
  end

  def unlock1
    @wants_to_enter1.make_false
  end

  def lock2
    @wants_to_enter2.make_true
    @turn = 1
    while @wants_to_enter1.true? && (@turn == 1)
      # Busy wait
    end
  end

  def unlock2
    @wants_to_enter2.make_false
  end
end

lock = MyLock.new
i = 0

thr1 = Thread.new do
  NUM_TIMES.times do
    lock.lock1
    i = i + 1
    lock.unlock1
  end
end

thr2 = Thread.new do
  NUM_TIMES.times do
    lock.lock2
    i = i + 1
    lock.unlock2
  end
end

thr1.join
thr2.join

puts i